numpy/lib/shape_base.py

```diff
@@ -16,9 +16,112 @@
 __all__ = [
     'column_stack', 'row_stack', 'dstack', 'array_split', 'split',
     'hsplit', 'vsplit', 'dsplit', 'apply_over_axes', 'expand_dims',
-    'apply_along_axis', 'kron', 'tile', 'get_array_wrap'
+    'apply_along_axis', 'kron', 'tile', 'get_array_wrap', 'take_along_axis'
     ]
 
+def take_along_axis(arr, indices, axis):
+    """
+    Take the elements described by `indices` along each 1-D slice of the given
+    `axis`, matching up subspaces of arr and indices.
+
+    This function can be used to index with the result of `argsort`, `argmax`,
+    and other functions.
+
+    This computes::
+
+        out[a..., k..., b...] = arr[a..., indices[a..., k..., b...], b...]
+
+    .. versionadded:: 1.13.0
+
+    Parameters
+    ----------
+    arr: array_like (A..., M, B...)
+        source array
+    indices: array_like (A..., K..., B...)
+        indices to take along each 1d slice of `arr`
+    axis: int
+        the axis to take 1d slices along
+
+    Returns
+    -------
+    out: ndarray (A..., K..., B...)
+        The indexed result.
+
+    See Also
+    --------
+    take : Take along an axis without matching up subspaces
+
+    Examples
+    --------
+
+    For this sample array
+
+    >>> a = np.array([[10, 30, 20], [60, 40, 50]])
+
+    We can sort either by using sort directly, or argsort and this function
+
+    >>> np.sort(a, axis=1)
+    array([[10, 20, 30],
+           [40, 50, 60]])
+    >>> ai = np.argsort(a, axis=1); ai
+    array([[0, 2, 1],
+           [1, 2, 0]], dtype=int64)
+    >>> np.take_along_axis(a, ai, axis=1)
+    array([[10, 20, 30],
+           [40, 50, 60]])
+
+    The same works for max and min:
+
+    >>> np.max(a, axis=1)
+    array([30, 60])
+    >>> ai = np.argmax(a, axis=1); ai
+    array([1, 0], dtype=int64)
+    >>> np.take_along_axis(a, ai, axis=1)
```

Comment:
> Without a usecase or thinking about it much more, I would be -0.5 on allowing nifty stuff.

I guess my argument here is the zero-one-infinity rule. We need to support indexing with zero and 1d arrays for `argmax` and `argsort` respectively, so we should just support all dimensions of arrays. One (contrived) example of supporting this would be:

```
x = np.rand(2, 3, 4)
axis = 1
xi1 = np.argsort(x, axis=axis )
xi2 = np.argsort(abs(x), axis=axis )

xis = np.stack([xi1, xi2], axis=1)
x_and_ax_sorted = np.take_along_axis(x, xis, axis=1)
# x_and_ax_sorted[i,0,:,j] = sorted(x[i,:,j])
# x_and_ax_sorted[i,1,:,j] = sorted(x[i,:,j], key=abs)
```

Without this feature, you'd need to call `take_along_axis` twice if you have two different `argsort` results.


In the same ways as you can call `expand_dims` before or after `take_along_axis`, supporting arbitrary dimensionality allows you to call `stack` before or after `take_along_axis`